### ISSUE TITLE
"RuntimeError: Given input size: (64x5x88). Calculated output size: (64x0x88). Output size is too small" when running vak train

### ISSUE THREAD

**Issue by marichard123** [NickleDave/vak#410]
I'm having a bit of an unexpected error- I receive the error "RuntimeError: Given input size: (64x5x88). Calculated output size: (64x0x88). Output size is too small" when running vak prep; full error traceback posted at the end. I had gotten vak running perfectly about two weeks ago, and today changed a few things from the previous run to apply vak to a larger dataset- I downloaded a larger dataset, modified my CSV annotation file pipeline code to be more efficient, and tweaked the training parameters in the config file. I then encountered the aforementioned error. I've made my best effort to restore running conditions to how they were when the program was running without issue, and as far as I can tell, conditions are identical- the bug still remains. The same bug also occurs when testing the program in previous virtual environments with vak installed that have not changed at all in the past few weeks.  I will continue to search for potential disparities between my current conditions and the conditions from two weeks ago, will attempt creating a new virtual environment and redownloading vak, and have requested that my partner attempt to run the program on his machine to see if he encounters the same bug.

```python
(vak-env5) PS C:\Users\Richard\documents\Fall_2021\Bat_Stuff\TweetynetPipeline> vak train .\gy6or6_train.toml
Logging results to C:\Users\Richard\Documents\Fall_2021\Bat_Stuff\TweetynetPipeline\results_211206_033123
Loading dataset from .csv path: C:\Users\Richard\Documents\Fall_2021\Bat_Stuff\TweetynetPipeline\TweetynetPipeline_prep_211206_032849.csv
Size of timebin in spectrograms from dataset, in seconds: 0.00128
using training dataset from C:\Users\Richard\Documents\Fall_2021\Bat_Stuff\TweetynetPipeline\TweetynetPipeline_prep_211206_032849.csv
Total duration of training split from dataset (in s): 50.01088
number of classes in labelmap: 2
will normalize spectrograms
Duration of WindowDataset used for training, in seconds: 50.01088000000001
Total duration of validation split from dataset (in s): 15.402240000000003
will measure error on validation set every 50 steps of training
Traceback (most recent call last):
  File "C:\Users\Richard\Anaconda3\envs\vak-env5\lib\site-packages\vak\models\models.py", line 69, in from_model_config_map
    model = MODELS[model_name].from_config(config=model_config)
KeyError: 'TweetyNet'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Richard\Anaconda3\envs\vak-env5\lib\runpy.py", line 192, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\Richard\Anaconda3\envs\vak-env5\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\Richard\Anaconda3\envs\vak-env5\Scripts\vak.exe\__main__.py", line 7, in <module>
  File "C:\Users\Richard\Anaconda3\envs\vak-env5\lib\site-packages\vak\__main__.py", line 45, in main
    cli.cli(command=args.command, config_file=args.configfile)
  File "C:\Users\Richard\Anaconda3\envs\vak-env5\lib\site-packages\vak\cli\cli.py", line 30, in cli
    COMMAND_FUNCTION_MAP[command](toml_path=config_file)
  File "C:\Users\Richard\Anaconda3\envs\vak-env5\lib\site-packages\vak\cli\train.py", line 52, in train
    core.train(
  File "C:\Users\Richard\Anaconda3\envs\vak-env5\lib\site-packages\vak\core\train.py", line 267, in train
    models_map = models.from_model_config_map(
  File "C:\Users\Richard\Anaconda3\envs\vak-env5\lib\site-packages\vak\models\models.py", line 71, in from_model_config_map
    model = MODELS[f"{model_name}Model"].from_config(
  File "C:\Users\Richard\Anaconda3\envs\vak-env5\lib\site-packages\tweetynet\model.py", line 15, in from_config
    network = TweetyNet(**config['network'])
  File "C:\Users\Richard\Anaconda3\envs\vak-env5\lib\site-packages\tweetynet\network.py", line 159, in __init__
    tmp_out = self.cnn(tmp_tensor)
  File "C:\Users\Richard\Anaconda3\envs\vak-env5\lib\site-packages\torch\nn\modules\module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "C:\Users\Richard\Anaconda3\envs\vak-env5\lib\site-packages\torch\nn\modules\container.py", line 117, in forward
    input = module(input)
  File "C:\Users\Richard\Anaconda3\envs\vak-env5\lib\site-packages\torch\nn\modules\module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "C:\Users\Richard\Anaconda3\envs\vak-env5\lib\site-packages\torch\nn\modules\pooling.py", line 153, in forward
    return F.max_pool2d(input, self.kernel_size, self.stride,
  File "C:\Users\Richard\Anaconda3\envs\vak-env5\lib\site-packages\torch\_jit_internal.py", line 267, in fn
    return if_false(*args, **kwargs)
  File "C:\Users\Richard\Anaconda3\envs\vak-env5\lib\site-packages\torch\nn\functional.py", line 585, in _max_pool2d
    return torch.max_pool2d(
RuntimeError: Given input size: (64x5x88). Calculated output size: (64x0x88). Output size is too small
(vak-env5) PS C:\Users\Richard\documents\Fall_2021\Bat_Stuff\TweetynetPipeline>
```

**Comment by marichard123** [comment]
Update: I have found that this bug, at least for now, seems unique to my machine: When running the exact same files and exact same commands on another computer, everything worked without issue- thus it seems that I accidentally made an undetected internal modification somewhere on my computer somehow.

To elaborate about the bug on my end:
On my computer, I had previously gotten a proof of concept test dataset working completely in Tweetynet- I subsequently downloaded a much larger quantity of data, changed some spectrogram parameters, and modified my data pipeline, attempting to improve prediction accuracy. After that- possibly because of my previously mentioned actions, possibly for any other reason- I encountered the bug. Even when I reset my files (pulling them out of the recycling bin) to to ones that I had used in the working proof of concept code, and completely redownload vak in a new virtual environment, the same problem occurred. 

On the other machine, I am working methodically to potentially isolate the source of the problem- if it occurs again, I will post an update along with the exact action that caused it.

**Comment by NickleDave** [comment]
Hi @marichard123 thank you for the update and thank you for the detail you provided above.

Are you still getting this error on the original machine?
```python
Traceback (most recent call last):
  File "C:\Users\Richard\Anaconda3\envs\vak-env5\lib\site-packages\vak\models\models.py", line 69, in from_model_config_map
    model = MODELS[model_name].from_config(config=model_config)
KeyError: 'TweetyNet'
```

That makes me wonder if you need to do `pip install tweetynet`

**Comment by NickleDave** [comment]
That might be a red herring though.

The other thing I can think of is that you just have a spectrogram that's literally too small for the network to process.

That's what this error makes me think:
```python
  File "C:\Users\Richard\Anaconda3\envs\vak-env5\lib\site-packages\torch\nn\functional.py", line 585, in _max_pool2d
    return torch.max_pool2d(
RuntimeError: Given input size: (64x5x88). Calculated output size: (64x0x88). Output size is too small
```

It looks like there's only 5 frequency bins in the input spectrogram?

**Comment by NickleDave** [comment]
You can do something like this to inspect the spectrogram sizes:
```python
import pandas as pd
import vak

vak_df = pandas.read_csv('your-csv-saved-by-vak-prep.csv')
spect_paths = vak_df.spect_path.values
sizes = [vak.files.spect.load(spect_path)['s'].shape for spect_path in spect_paths]  # load each spectrogram and get shape
print(sorted(sizes, key=lambda x: x[0])[:10])  # now sort by first dimension which is frequency bins. Only print 10 smallest
```

Not sure if that exact snippet works, I'm not at a terminal to test, but hopefully you see what I'm trying to do

**Comment by marichard123** [comment]
Hello @NickleDave-
I am currently trying to run the piece of diagnostic code, but am running into bugs. I have simplified it to 
```python
import pandas as pd
import vak
vak_df = pd.read_csv('C:/Users/Richard/Documents/Fall_2021/Bat_Stuff/TweetynetPipeline/PipelineCSVOutput.csv')
spect_paths = 'C:/Users/Richard/Documents/Fall_2021/Bat_Stuff/TweetynetPipeline/spectrograms_generated_211208_035107/Logger3_14_200123_0958_VocExtractData1.wav.spect'
size = vak.files.spect.load(spect_paths)['s'].shape
print(size)
```
but am running into a bug, which I have posted at the very end. **However, I believe I have isolated the cause of the problem- I am almost certain that changing the spectrogram parameter values in the training and prediction toml configuration files is the issue.** On the other machine which I had been using since my own computer began experiencing bugs, I had been changing small things such as the size of the dataset or the pipeline that I was using in an attempt to isolate the problem. After I changed the value of the fft_size from the default 512 to 476. I then experienced the following error:
```python
(vak-env) PS C:\Users\tobias\desktop\Tweetynet> vak prep gy6or6_train.toml
determined that purpose of config file is: train
will add 'csv_path' option to 'TRAIN' section
purpose for dataset: train
will split dataset
making array files containing spectrograms from audio files in: C:\Users\tobias\Desktop\Tweetynet
Traceback (most recent call last):
  File "C:\Users\tobias\anaconda3\envs\vak-env\lib\site-packages\vak\annotation.py", line 232, in source_annot_map
    annot = audio_stem_annot_map[source_file_stem]
KeyError: 'Logger10_78_200123_0958_VocExtractData4'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "C:\Users\tobias\anaconda3\envs\vak-env\lib\runpy.py", line 192, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\tobias\anaconda3\envs\vak-env\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\tobias\anaconda3\envs\vak-env\Scripts\vak.exe\__main__.py", line 7, in <module>
  File "C:\Users\tobias\anaconda3\envs\vak-env\lib\site-packages\vak\__main__.py", line 45, in main
    cli.cli(command=args.command, config_file=args.configfile)
  File "C:\Users\tobias\anaconda3\envs\vak-env\lib\site-packages\vak\cli\cli.py", line 30, in cli
    COMMAND_FUNCTION_MAP[command](toml_path=config_file)
  File "C:\Users\tobias\anaconda3\envs\vak-env\lib\site-packages\vak\cli\prep.py", line 132, in prep
    vak_df, csv_path = core.prep(
  File "C:\Users\tobias\anaconda3\envs\vak-env\lib\site-packages\vak\core\prep.py", line 205, in prep
    vak_df = dataframe.from_files(
  File "C:\Users\tobias\anaconda3\envs\vak-env\lib\site-packages\vak\io\dataframe.py", line 133, in from_files
    spect_files = audio.to_spect(
  File "C:\Users\tobias\anaconda3\envs\vak-env\lib\site-packages\vak\io\audio.py", line 174, in to_spect
    audio_annot_map = source_annot_map(audio_files, annot_list)
  File "C:\Users\tobias\anaconda3\envs\vak-env\lib\site-packages\vak\annotation.py", line 234, in source_annot_map
    raise ValueError(
ValueError: could not find annotation for source file: C:\Users\tobias\Desktop\Tweetynet\Logger10_78_200123_0958_VocExtractData4.wav.
No annotation had an audio file whose stem matched the source file stem: Logger10_78_200123_0958_VocExtractData4
```
I had, however, made absolutely no changes to the audio files or the CSV annotation file, and upon manually checking, the audio file and its reference were indeed reference. Upon changing the value of fft_size back to the original value of 512, the bug persisted. As far as I can tell, by making that change, I put myself in the exact same situation as with my own computer. I had in fact experienced this same bug on my own computer when messing around with various configurations of data in an attempt to solve the original problem. At the time, I had assumed that I myself had deleted some files and began trying other solutions. I have not yet the bug on the other machine much yet, but I will try various combinations to see if the error traceback changes. However, from the fact that on my own machine not even redownloading vak and restoring old previously working data files fixed the issue, I hypothesize that some unexpected internal fundamental change occurred when I changed the parameter value.




The previously mentioned error I got when trying to get spectrogram size:
```python
(vak-env6) PS C:\Users\Richard\Documents\Fall_2021\Bat_stuff> python RandomStuff2.py
Traceback (most recent call last):
  File "RandomStuff2.py", line 16, in <module>
    size = vak.files.load(spect_paths)['s'].shape
AttributeError: module 'vak.files' has no attribute 'load'
(vak-env6) PS C:\Users\Richard\Documents\Fall_2021\Bat_stuff> python RandomStuff2.py
Traceback (most recent call last):
  File "RandomStuff2.py", line 16, in <module>
    size = vak.files.spect.load(spect_paths)['s'].shape
  File "C:\Users\Richard\Anaconda3\envs\vak-env6\lib\site-packages\vak\files\spect.py", line 84, in load
    spect_dict = constants.SPECT_FORMAT_LOAD_FUNCTION_MAP[spect_format](spect_path)
KeyError: 'spect'
(vak-env6) PS C:\Users\Richard\Documents\Fall_2021\Bat_stuff>
```

**Comment by NickleDave** [comment]
Hi @marichard123 thank you for providing these details and the complete tracebacks.

I can't think of anything `prep` would do that could somehow change its ability to find your data.

Maybe this is a case where it would be better to try a zoom meeting for tech support? 
I think I'd be able to get a better sense of what's going on if I help you troubleshoot in real time.

**Comment by NickleDave** [comment]
@marichard123 I'm going to close this for now after we were able to get things running with the defaults.
But just let me know if you run into any more hitches

for posterity: part of the issue was due to the way the `vak.spect.spectrogram` function sets frequency cutoffs.
We may need to add some functionality to it.
@julieelie suggested the approach in `soundsig` might be better, where there's a `freq_spacing` parameter.
https://github.com/theunissenlab/soundsig/blob/fca413aa71ce6fec079c59e615e328e6781d1d35/soundsig/sound.py#L103
Not clear to me where that `spectrogram` functions come from in the code at the moment.
The basic issue is that we are generating the spectrogram from the full signal even after the bandpass filter, producing a bunch of frequency bins with essentially no power.

I think we would need to somehow down sample so we can still have a good frequency resolution?
Something like this:
https://www.dsprelated.com/thread/5310/understanding-subband-filtering

Currently feeling like I do not want to be in the business of writing a DSP library though, and that we should just instead have a very helpful message saying, "here is how to supply your own spectrograms" (as in #413)
